### DISCUSSION TITLE
Test-DbaBuild Not Pulling cu22 for SQL 2019 Server

### DISCUSSION THREAD

**Discussion by sueb-lex** [dataplat/dbatools#9067] [General] [answered]
When running below command it does not pull the latest cu22 for SQL 2019.  Server has cu21 and command shows it is compliant.  I am using the latest dbatools version 2.0.4.

Test-DbaBuild -SqlInstance Sql2019Server -MaxBehind '0CU'

**Comment by niphlod** [comment]
did you try updating the index ? https://docs.dbatools.io/Update-DbaBuildReference

**Comment by sueb-lex** [comment]
I ran the Update-DbaBuildReference and then re-ran Test-DbaBuild and there is still an issue.

***@***.***

**Reply by sueb-lex** [reply]
Still only shows the cu21 as the latest for SQL 2019 and does not pick up the cu22.

**Reply by niphlod** [reply]
what if you do Test-DbaBuild -Build "15.0.4322" ?

**Comment by sueb-lex** [comment]
Had to run below for it to work (added -Latest):

Test-DbaBuild -Build "15.0.4322" -Latest

Here is the error I receive
WARNING: [10:03:48][Test-DbaBuild] 15.0.4322 is not recognized as a correct version

**Reply by niphlod** [reply]
then I'm not sure you successfully updated the index .
Even with a reaaaly old index, this is what happens ... (if you do -Update on Test-DbaBuild it'll try to do Update-DbaBuildReference)

with an old index
![image](https://github.com/dataplat/dbatools/assets/122119/217e7e58-e231-4a75-a59f-79a561d65a72)
and then
![image](https://github.com/dataplat/dbatools/assets/122119/633b1f57-5cb3-41ad-85aa-80b76d045690)

**Reply by sueb-lex** [reply]
Ran this command and this is what I get.  When updating dbatools to 2.0.4 has to run command below to get it to work.  Could that have anything to do with this issue?

Set-DbatoolsInsecureConnection

![image](https://github.com/dataplat/dbatools/assets/9084717/5622c3d9-42b8-45fe-90a8-fc7ab180cbc6)

**Reply by sueb-lex** [reply]
Now every server that I have applied SQL2019 cu22 to gives this error message when trying to check the current build.  No idea how to fix this issue.

WARNING: [13:11:36][Test-DbaBuild] 15.0.4322 is not recognized as a correct version

**Reply by niphlod** [reply]
once again, until you're able to correctly pull down the updated build reference it's not going to show anything, because it seems that your current reference doesn't hold details regarding 15.0.4322

**Reply by rvsuresh19** [reply]
Hi niphlod,

I'm also facing similar problem as reported originally, but this issue is not faced for sql server 2022. I have followed the instructions that you have shared but still facing issue. Can you please review the screenshots and guide me if I'm doing anything wrong here?

_Issue with SQL 2019 latest patch_
**test-dbabuild -Latest -sqlinstance myinstancename**
![image](https://github.com/dataplat/dbatools/assets/54534997/76c6ef4e-1b9e-47a6-9528-9cb9c277c4ac)


**test-dbabuild -Build "15.0.4322" -Latest -Update -sqlinstance myinstancename**
![image](https://github.com/dataplat/dbatools/assets/54534997/29144ddd-b1b4-4550-a7db-889173e324f6)

**test-dbabuild -Build "15.0.4322" -sqlinstance myinstancename -MaxBehind '0CU'**
![image](https://github.com/dataplat/dbatools/assets/54534997/a95fa7ee-217f-47a0-9ec1-38f34010297e)

_No issues with SQL 2022 patches_
**test-dbabuild -Latest -sqlinstance myinstancename22**
![image](https://github.com/dataplat/dbatools/assets/54534997/ddd2d584-8c9a-405c-8ae2-213cc1f509be)


Version
![image](https://github.com/dataplat/dbatools/assets/54534997/690e77ea-37a1-440b-b664-338e858fe73c)

**Reply by niphlod** [reply]
let's go back one sec. Test-DbaBuild either works detecting the version from the -SqlInstance or the -Build . So passing both in doesn't make much sense.
Second, Test-DbaBuild works with data contained in a json that we'll call "reference" . That json file is the same one that https://dataplat.github.io/builds uses to show all the versions.
Third, you *may* have an old dbatools version, and still be able to detect "recent" version, as long as the "reference" is kept updated.
To do so, you can do 2 things: run Test-DbaBuild once with -Update, or run Update-DbaBuildReference. Both will fetch from internet the new "referece" json so Test-DbaBuild can work with the most recent data available (which BTW is just a long list of versions annotated with the corresponding releases such as SPs and CUs).
The topic/discussion revolves around an incorrect detection of 2019 cu22. To make sure you have what ATM is the latest data, you run Update-DbaBuildReference. Then, run `Test-DbaBuild -Build "15.0.4322"` (which is 2019 cu22). If the result matchtype is "exact", then Test-DbaBuild is working as expected and you have the latest data. If matchtype is approximate, your "reference" is stale.
Till you have a working version with the latest data available, debating on any other issue regarding version comparison (such as -MaxBehind) is not useful.

**Reply by rvsuresh19** [reply]
Hi ,

Thank you very much for taking time and explaining it in detail! I understand my stale "reference" is the root cause for not able to read latest 2019 CU. I tried running Update-DbaBuildReference & even Test-DbaBuild with -Update, but it is not helping the latest reference to get loaded 

![image](https://github.com/dataplat/dbatools/assets/54534997/5532a324-7dff-463f-94ca-e4ed99ecc388)

I checked the json file(dbatools-buildref-index), interestingly 2019 had entries only until CU20..even CU21 didn't existed, but unsure how it detects CU21

![image](https://github.com/dataplat/dbatools/assets/54534997/6b96ac2a-38ea-4759-9d4c-bd146b3238f3)

So I tried adding the entries manually for both 21 & 22
![image](https://github.com/dataplat/dbatools/assets/54534997/99ad27ad-fc82-45b2-b5a0-61d374e35151)


and hoping this would work, but unfortunately it didn't. Can you please advise if there is something I can try to get it sorted?

**Reply by niphlod** [reply]
what if you run `Update-DbaBuildReference -Verbose` ?

**Reply by rvsuresh19** [reply]
![image](https://github.com/dataplat/dbatools/assets/54534997/e10ab2ff-ae62-4777-be83-4fba1c1c17a1)

**Reply by niphlod** [reply]
then in Get-DbatoolsConfigValue -Name 'Path.DbatoolsData' you should have an updated dbatools-buildref-index.json .... or something smells very fishy. If it isn't able to pull the index it would have spitted out a warning message.

**Reply by rvsuresh19** [reply, marked as answer]
Thank you very much that sorted the problem. I had a file residing inside user profile AppData roaming folder. I tried removing the existing json file from the folder and executed the Update-DbaBuildReference and it fetched the latest json with all CUs and issue got resolved. Thanks again :)

![image](https://github.com/dataplat/dbatools/assets/54534997/e20ad5e8-e392-4d31-bf22-016c96b86ea1)

**Reply by niphlod** [reply]
normally it would compare your local version with the index pulled from internet and if the internet one is newer, proceed to update it. Seems like for some reason your local one was detected as newer than the one pulled from internet (for comparison it uses the "LastUpdated" property at the very top of the file)

**Comment by sueb-lex** [comment]
I have upgraded dbatools on an different server (Server 2) to 1.1.146 and it works as expected.  

I have removed v. 2.0.4 on Server 1 and it still will not work.  Whatever I have tested on Server 1 (uninstall new version, re-install new version, set to trust cert) I am not able to get Server 1 to update to SQL 2019 cu22.  Not sure what else to try.

**Reply by sueb-lex** [reply]
I located the json file as suggested by niphlod and renamed it.  Then followed his steps.  This solved my issue as well.

Thank you all so much for helping to resolve this.

**Reply by wsmelton** [reply]
Can you provide that renamed file @sueb-lex ? We'd like to check why it chose not to pull an updated version. (see if we can replicate the problem)

**Reply by sueb-lex** [reply]
![image](https://github.com/dataplat/dbatools/assets/9084717/4b2ea6e5-d1af-44c4-9c43-e9aa86e03277)

**Reply by wsmelton** [reply]
Can you attach that file to this discussion?

**Reply by sueb-lex** [reply]
[dbatools-buildref-index.zip](https://github.com/dataplat/dbatools/files/12488170/dbatools-buildref-index.zip)
